commit 72d43d9bc9210d24d09202eaf219eac09e17b339
Author: Thomas Gleixner <tglx@linutronix.de>
Date:   Sat Sep 6 03:06:08 2008 +0200

    x86: HPET: read back compare register before reading counter
    
    After fixing the u32 thinko I sill had occasional hickups on ATI chipsets
    with small deltas. There seems to be a delay between writing the compare
    register and the transffer to the internal register which triggers the
    interrupt. Reading back the value makes sure, that it hit the internal
    match register befor we compare against the counter value.
    
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

---
 arch/x86/kernel/hpet.c |    7 +++++++
 1 file changed, 7 insertions(+)

Index: linux-2.6.24.7-rt21/arch/x86/kernel/hpet.c
===================================================================
--- linux-2.6.24.7-rt21.orig/arch/x86/kernel/hpet.c	2008-10-08 22:25:17.000000000 -0400
+++ linux-2.6.24.7-rt21/arch/x86/kernel/hpet.c	2008-10-08 22:25:17.000000000 -0400
@@ -289,6 +289,13 @@ static int hpet_legacy_next_event(unsign
 	cnt += (u32) delta;
 	hpet_writel(cnt, HPET_T0_CMP);
 
+	/*
+	 * We need to read back the CMP register to make sure that
+	 * what we wrote hit the chip before we compare it to the
+	 * counter.
+	 */
+	WARN_ON((u32)hpet_readl(HPET_T0_CMP) != cnt);
+
 	return (s32)((u32)hpet_readl(HPET_COUNTER) - cnt) >= 0 ? -ETIME : 0;
 }
 
